From 71d94fed46f9c42c9d8207fbd166f625e1c1a84c Mon Sep 17 00:00:00 2001
From: "Christopher A. Williamson" <contact@chrisaw.io>
Date: Sun, 24 May 2026 17:34:18 +0100
Subject: [PATCH] dnsmasq: add verbose debug logging for startup and pipe
 communication

Add DNSMASQ_DEBUG log lines at every startup milestone so we can
see exactly how far dnsmasq gets before exiting. Also log when
the stdin pipe is closed by peer (netd) and log every fatal_event.

This is debug-builds only.

Issue: #14
Change-Id: I4baedb3e6b77d6dca7219d8007f3af8b0ce0b02b
---
 src/dnsmasq.c | 24 +++++++++++++++++++++++-
 src/log.c     |  1 +
 2 files changed, 24 insertions(+), 1 deletion(-)

diff --git a/src/dnsmasq.c b/src/dnsmasq.c
index 46d15ee..78168bb 100644
--- a/src/dnsmasq.c
+++ b/src/dnsmasq.c
@@ -147,10 +147,14 @@ int main(int argc, char** argv) {
 #endif
 
     setupSignalHandling();
+    my_syslog(LOG_INFO, "DNSMASQ_DEBUG: signals setup done");
 
     umask(022); /* known umask, create leases and pid files as 0644 */
 
     read_opts(argc, argv, compile_opts);
+    my_syslog(LOG_INFO, "DNSMASQ_DEBUG: opts parsed, port=%d user=%s group=%s",
+              daemon->port, daemon->username ? daemon->username : "(null)",
+              daemon->groupname ? daemon->groupname : "(null)");
 
     if (daemon->edns_pktsz < PACKETSZ) daemon->edns_pktsz = PACKETSZ;
     daemon->packet_buff_sz =
@@ -164,6 +168,7 @@ int main(int argc, char** argv) {
 #endif
 
     closeUnwantedFileDescriptors();
+    my_syslog(LOG_INFO, "DNSMASQ_DEBUG: unwanted fds closed");
 
 #ifdef HAVE_LINUX_NETWORK
     netlink_init();
@@ -187,6 +192,7 @@ int main(int argc, char** argv) {
         dhcp_init();
     }
 #endif
+    my_syslog(LOG_INFO, "DNSMASQ_DEBUG: dhcp init done");
 
     if (!enumerate_interfaces()) die(_("failed to find list of interfaces: %s"), NULL, EC_MISC);
 
@@ -205,10 +211,13 @@ int main(int argc, char** argv) {
     } else if ((daemon->port != 0 || (daemon->options & OPT_TFTP)) &&
                !(daemon->listeners = create_wildcard_listeners()))
         die(_("failed to create listening socket: %s"), NULL, EC_BADNET);
+    my_syslog(LOG_INFO, "DNSMASQ_DEBUG: listeners created, port=%d listeners=%p",
+              daemon->port, (void*)daemon->listeners);
 
     if (daemon->port != 0) cache_init();
 
     if (daemon->port != 0) pre_allocate_sfds();
+    my_syslog(LOG_INFO, "DNSMASQ_DEBUG: cache and sfds ready");
 
 #if defined(HAVE_DHCP) && defined(HAVE_SCRIPT)
     /* Note getpwnam returns static storage */
@@ -225,6 +234,9 @@ int main(int argc, char** argv) {
         baduser = daemon->username;
     else if (daemon->groupname && !(gp = getgrnam(daemon->groupname)))
         baduser = daemon->groupname;
+    my_syslog(LOG_INFO, "DNSMASQ_DEBUG: user lookup done, username=%s baduser=%s ent_pw=%p",
+              daemon->username ? daemon->username : "(null)",
+              baduser ? baduser : "(null)", (void*)ent_pw);
 
     if (baduser) die(_("unknown user or group: %s"), baduser, EC_BADCONF);
 
@@ -601,6 +613,8 @@ void send_event(int fd, int event, int data) {
     ev.event = event;
     ev.data = data;
 
+    my_syslog(LOG_INFO, "DNSMASQ_DEBUG: send_event(event=%d, data=%d, fd=%d)", event, data, fd);
+
     /* error pipe, debug mode. */
     if (fd == -1)
         fatal_event(&ev);
@@ -614,6 +628,9 @@ void send_event(int fd, int event, int data) {
 static void fatal_event(struct event_desc* ev) {
     errno = ev->data;
 
+    my_syslog(LOG_ERR, "DNSMASQ_DEBUG: fatal_event(event=%d, data=%d, errno=%d)",
+              ev->event, ev->data, errno);
+
     switch (ev->event) {
         case EVENT_DIE:
             exit(0);
@@ -808,10 +825,15 @@ static int check_android_listeners(fd_set* set) {
         int consumed = 0;
 
         if ((rc = read(STDIN_FILENO, buffer, sizeof(buffer) - 1)) < 0) {
-            my_syslog(LOG_ERR, _("Error reading from stdin (%s)"), strerror(errno));
+            my_syslog(LOG_ERR, "DNSMASQ_DEBUG: Error reading from stdin (%s)", strerror(errno));
+            return -1;
+        }
+        if (rc == 0) {
+            my_syslog(LOG_ERR, "DNSMASQ_DEBUG: read(STDIN) returned 0, pipe closed by peer (netd)");
             return -1;
         }
         buffer[rc] = '\0';
+        my_syslog(LOG_INFO, "DNSMASQ_DEBUG: read %d bytes from stdin/netd", rc);
         while (consumed < rc) {
             char* cmd;
             char* current_cmd = &buffer[consumed];
diff --git a/src/log.c b/src/log.c
index 1cc4745..7efdff4 100644
--- a/src/log.c
+++ b/src/log.c
@@ -377,6 +377,7 @@ void die(char* message, char* arg1, int exit_code) {
     log_stderr = 0;
     my_syslog(LOG_CRIT, _("FAILED to start up"));
     flush_log();
+    my_syslog(LOG_CRIT, "DNSMASQ_DEBUG: die() called, exit_code=%d", exit_code);
 
     exit(exit_code);
 }
-- 
2.54.0

